### PR TITLE
athena audit logs - add metrics

### PR DESCRIPTION
**Pull request by tobiaszheller**
Part of https://github.com/gravitational/teleport.e/issues/894
RFD: https://github.com/gravitational/teleport/blob/master/rfd/0118-scalable-audit-logs.md

This PR adds metrics to athena logger.

### CODE REVIEW

**Comment by tobiaszheller** [comment]
@evanfreed @logand22 I would like your feedback also on metrics used here. Feel free to add comment if you think something is missing or not right.

File: lib/events/athena/consumer.go
Comment (smallinsky) [inline lib/events/athena/consumer.go:304]: Do we really need do it async ? 
It just a few inc/dec/ if calls that should be insignificant in terms of performance.

File: lib/events/athena/consumer.go
Comment (smallinsky) [inline lib/events/athena/consumer.go:668]: nit: 
What would you say about moving metrics logic into `sqsMessagesCollector.fromSQS` instead of creating a artificial pipeline ?

File: lib/events/athena/consumer.go
Comment (smallinsky) [inline lib/events/athena/consumer.go:288]: Just to ensure. Do we want to include events collection time to the  "BatchProcessing duration" ?

File: lib/events/athena/consumer.go
Comment (tobiaszheller) [reply lib/events/athena/consumer.go]: I initially started with few inc/dec/ if calls but was not happy with readability. I give it another try in https://github.com/gravitational/teleport/pull/26331/commits/29998bbc2eb23c38a06621fb72e7c666ba8e3320 and I think now it's more readable. I have dropped that sync solution.

File: lib/events/athena/consumer.go
Comment (tobiaszheller) [reply lib/events/athena/consumer.go:289]: yes

File: lib/events/athena/athena.go
Comment (tobiaszheller) [reply lib/events/athena/athena.go:435]: reworked in https://github.com/gravitational/teleport/pull/26331/commits/9c6211e0aee1b9a1f3db9fb9232649f255174543

File: lib/events/athena/athena.go
Comment (tobiaszheller) [reply lib/events/athena/athena.go:441]: if batcher will be set to 1min (60s), those values represents buckets that I think will be interested. 

if we want to use fn(batchInterval), I was thinking about something like:
```
0.1 * batchInterval, 0.2 * batchInterval, 0.5 * batchInterval, 0.75 * batchInterval, prometheus.ExponentialBucketsRange(0.9*batchInterval, 2*batchInterval, 10)
```
which evaluates into:
```
6.00
12.00
30.00
45.00
54.00
59.01
64.48
70.47
77.01
84.15
91.96
100.49
109.81
120.00
```

do you have better suggestion for those bucket values?

**Review by smallinsky — APPROVED** [review]
LGTM once remaining comments are addressed.

File: lib/events/athena/athena.go
Comment (tobiaszheller) [reply lib/events/athena/athena.go]: ah, I see, reworked in https://github.com/gravitational/teleport/pull/26331/commits/2b20402d09a38634183cc42d0a87c7895fa7a723.

File: lib/events/athena/athena.go
Comment (tobiaszheller) [reply lib/events/athena/athena.go]: in https://github.com/gravitational/teleport/pull/26331/commits/2b20402d09a38634183cc42d0a87c7895fa7a723 i have reworked that metric to be fn of batchInterval and used values proposed in comment above. We will have some visibility if batch is very fast, but mostly we are interested in range 0.9* batch interval to 2* batch interval.

**Review by rosstimothy — APPROVED** [review]

File: docs/pages/includes/metrics.mdx
Comment (zmb3) [inline docs/pages/includes/metrics.mdx:21]: The `|` on lines 21 and 22 should be moved further left.

File: docs/pages/includes/metrics.mdx
Comment (zmb3) [inline docs/pages/includes/metrics.mdx:52]: Metric names are hard to change because people rely on them once they have dashboards, alerts, etc.

Should we try to come up with a name that's not specific to Athena or S3 (so that when we get this working on prem with alternate technologies the metric names still make sense)?

File: docs/pages/includes/metrics.mdx
Comment (alexfornuto) [reply docs/pages/includes/metrics.mdx:21]: I could just run this whole thing through a markdown table formatter if the author is OK with it.

Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx]: fixed in https://github.com/gravitational/teleport/pull/26331/commits/daf76446215db2fc7ce2f7088ec65a3aa7f41aad

File: docs/pages/includes/metrics.mdx
Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx:52]: for prefix we can use `teleport_audit_presto_`? or `teleport_audit_objectstorage_`? 

or we can go with more generic names like: `teleport_audit_batch_processing_time` and `teleport_audit_data_flush_time`, and use description here to indicate that so far they are used only in athena based logger. What do you think @zmb3 ?

Comment (zmb3) [reply docs/pages/includes/metrics.mdx:52]: That works for me.

Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx:52]: @rosstimothy  @zmb3 right now i think that going that generic as `teleport_audit_data_flush_time` may not be best option. You can configure multiple events logger, and if we go with very generic name like `teleport_audit_data_flush_time` and someone uses two loggers that use the same metrics, data will be invalid because both loggers will write to the same metric. (because they are not vector labels).

So what we can do is either introduce that new generic metric as vector with label as 'athena' for now, or go with `teleport_audit_presto_` or keep `teleport_audit_athena_`.

Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx:52]: @zmb3 @rosstimothy ☝️

Comment (zmb3) [reply docs/pages/includes/metrics.mdx:52]: I don't have a strong preference and leave it up to you to decide on names that make the most sense. You are more familiar with the subject matter than I am :-)

Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx]: i went with `parquetlog`

**Comment by tobiaszheller** [comment]
@alexfornuto @ptgott friendly ping :)

File: docs/pages/includes/metrics.mdx
Comment (ptgott) [inline docs/pages/includes/metrics.mdx:2]: Is there a way to reduce the size of the diff in this partial? I'm not really sure what content this PR adds here.

Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx:2]: @ptgott you can use hide whitespaces changes in diff 
Screenshot 2023-05-25 at 08 50 06

with that settings you can see that there are added only new metrics and formatting update due to size of first colum:
```
| `teleport_audit_parquetlog_batch_processing_seconds`     | histogram | Teleport Audit Log  | Duration of processing single batch of events in parquet audit log.                                |
| `teleport_audit_parquetlog_s3_flush_seconds`             | histogram | Teleport Audit Log  | Duration of flushing parquet files to s3 in parquet audit log.                                     |
| `teleport_audit_parquetlog_delete_events_seconds`        | histogram | Teleport Audit Log  | Duration of deletion events from SQS in parquet audit log.                                         |
| `teleport_audit_parquetlog_batch_size`                   | histogram | Teleport Audit Log  | Overall size of events in single batch in parquet audit log.                                       |
| `teleport_audit_parquetlog_batch_count`                  | counter   | Teleport Audit Log  | Total number of events in single batch in parquet audit log.                                       |
| `teleport_audit_parquetlog_last_processed_timestamp`     | gauge     | Teleport Audit Log  | Number of last processing time in parquet audit log.                                               |
| `teleport_audit_parquetlog_age_oldest_processed_message` | gauge     | Teleport Audit Log  | Number of age of oldest event in parquet audit log.                                                |
| `teleport_audit_parquetlog_errors_from_collect_count`    | counter   | Teleport Audit Log  | Number of collect failures in parquet audit log.    
```

File: docs/pages/includes/metrics.mdx
Comment (ptgott) [inline docs/pages/includes/metrics.mdx:51]: Will readers of the docs know what "parquet audit log" means? "parquet" isn't mentioned in other places in the docs. 

Is there a plan to add instructions for configuring the Parquet audit log? Otherwise, I would make this something like, "Duration of processing a single batch of events in the Parquet-format audit log, if this is enabled".

This comment applies to the other new metric descriptions as well.

File: docs/pages/includes/metrics.mdx
Comment (ptgott) [inline docs/pages/includes/metrics.mdx:52]: We should capitalize third-party product names as they are capitalized in their docs. This means using "Parquet", "S3", etc.

File: docs/pages/includes/metrics.mdx
Comment (tobiaszheller) [reply docs/pages/includes/metrics.mdx]: yes, in separate PR I will add configuration of Parquet based audit log. I have updated description, but omitted "if this is enabled" because we don't use if for any other audit log or backend. PTAL.

**Comment by tobiaszheller** [comment]
@ptgott I have applied you comments, PTAL

**Review by ptgott — APPROVED** [review]

**Pull request merged**
